File: nebras-dashboard/app/src/main/kotlin/com/nebras/dashboard/data/StorageUpload.kt
package com.nebras.dashboard.data

import android.net.Uri
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageException
import com.google.firebase.storage.StorageMetadata
import com.google.firebase.storage.UploadTask
import kotlinx.coroutines.tasks.await
import java.io.File

/** نتيجة الرفع — نظير `SmartUploadResult` في `smartUpload.js`. */
data class UploadResult(
    val url: String,
    val path: String,
    val contentType: String,
    val size: Long,
    val filename: String,
)

private val UNSAFE_SEGMENT = Regex("[#$\\[\\]./\\\\:*?\"<>|]+")
private val FOLDER_EDGES = Regex("^/+|/+$")

private fun sanitizeSegment(name: String): String {
    // إزالة أحرف التحكّم (0x00-0x1F و 0x7F) دون regex لتفادي مشاكل الترميز.
    val buffer = StringBuilder()
    for (ch in name.trim()) {
        val cu = ch.code
        if (cu <= 0x1F || cu == 0x7F) continue
        buffer.append(ch)
    }
    var s = buffer.toString()
    s = UNSAFE_SEGMENT.replace(s, "_")
    if (s.length > 180) s = s.substring(0, 180)
    return if (s.isEmpty()) "file" else s
}

/**
 * رفع بايتات إلى دلو Nebras (مثل `smartUpload` → Storage). الترتيب الثابت:
 * Storage أولاً ثم نُعيد رابط التنزيل ليُكتب في Firestore.
 *
 * [onProgress] نسبة مئويّة 0..100. [isAborted] لإلغاء الرفع.
 */
suspend fun smartUpload(
    bytes: ByteArray,
    filename: String,
    contentType: String = "application/octet-stream",
    folder: String = "dashboard/uploads",
    onProgress: ((Double) -> Unit)? = null,
    isAborted: (() -> Boolean)? = null,
    onTaskCreated: ((UploadTask) -> Unit)? = null,
): UploadResult {
    val storage = FirebaseStorage.getInstance()
    val safeName = sanitizeSegment(filename)
    val cleanFolder = FOLDER_EDGES.replace(folder, "")
    val path = "$cleanFolder/${System.currentTimeMillis()}_$safeName"
    val ref = storage.getReference(path)
    val ct = if (contentType.trim().isEmpty()) "application/octet-stream" else contentType

    val task = ref.putBytes(bytes, StorageMetadata.Builder().setContentType(ct).build())
    onTaskCreated?.invoke(task)

    task.addOnProgressListener { snap ->
        if (isAborted != null && isAborted()) {
            task.cancel()
            return@addOnProgressListener
        }
        val total = if (snap.totalByteCount <= 0) bytes.size.toLong() else snap.totalByteCount
        if (total > 0) onProgress?.invoke((snap.bytesTransferred.toDouble() / total) * 100.0)
    }

    try {
        task.await()
    } catch (e: StorageException) {
        if (e.errorCode == StorageException.ERROR_CANCELED) {
            throw IllegalStateException("تمّ إلغاء الرفع.")
        }
        throw e
    }

    if (isAborted != null && isAborted()) {
        throw IllegalStateException("تمّ إلغاء الرفع.")
    }

    val url = ref.downloadUrl.await().toString()
    return UploadResult(
        url = url,
        path = path,
        contentType = ct,
        size = bytes.size.toLong(),
        filename = safeName,
    )
}

/**
 * رفع ملفّ من المسار (بثّ — لا يُحمَّل كاملاً في الذاكرة). يُستعمل للملفّات
 * الكبيرة في الرفع المتعدّد لتفادي OOM.
 */
suspend fun smartUploadFile(
    file: File,
    filename: String,
    contentType: String = "application/octet-stream",
    folder: String = "dashboard/uploads",
    onProgress: ((Double) -> Unit)? = null,
    isAborted: (() -> Boolean)? = null,
    onTaskCreated: ((UploadTask) -> Unit)? = null,
): UploadResult {
    val storage = FirebaseStorage.getInstance()
    val safeName = sanitizeSegment(filename)
    val cleanFolder = FOLDER_EDGES.replace(folder, "")
    val path = "$cleanFolder/${System.currentTimeMillis()}_$safeName"
    val ref = storage.getReference(path)
    val ct = if (contentType.trim().isEmpty()) "application/octet-stream" else contentType
    val size = file.length()

    val task = ref.putFile(
        Uri.fromFile(file),
        StorageMetadata.Builder().setContentType(ct).build(),
    )
    onTaskCreated?.invoke(task)

    task.addOnProgressListener { snap ->
        if (isAborted != null && isAborted()) {
            task.cancel()
            return@addOnProgressListener
        }
        val total = if (snap.totalByteCount <= 0) size else snap.totalByteCount
        if (total > 0) onProgress?.invoke((snap.bytesTransferred.toDouble() / total) * 100.0)
    }

    try {
        task.await()
    } catch (e: StorageException) {
        if (e.errorCode == StorageException.ERROR_CANCELED) {
            throw IllegalStateException("تمّ إلغاء الرفع.")
        }
        throw e
    }

    if (isAborted != null && isAborted()) {
        throw IllegalStateException("تمّ إلغاء الرفع.")
    }

    val url = ref.downloadUrl.await().toString()
    return UploadResult(
        url = url,
        path = path,
        contentType = ct,
        size = size,
        filename = safeName,
    )
}

File: dashboard-flutter/lib/data/storage_upload.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:firebase_storage/firebase_storage.dart';

/// نتيجة الرفع — نظير `SmartUploadResult` في `smartUpload.js`.
class UploadResult {
  UploadResult({
    required this.url,
    required this.path,
    required this.contentType,
    required this.size,
    required this.filename,
  });

  final String url;
  final String path;
  final String contentType;
  final int size;
  final String filename;
}

String _sanitizeSegment(String name) {
  // إزالة أحرف التحكّم (0x00-0x1F و 0x7F) دون regex لتفادي مشاكل الترميز.
  final buffer = StringBuffer();
  for (final cu in name.trim().codeUnits) {
    if (cu <= 0x1F || cu == 0x7F) continue;
    buffer.writeCharCode(cu);
  }
  var s = buffer.toString();
  s = s.replaceAll(RegExp(r'[#$\[\]./\\:*?"<>|]+'), '_');
  if (s.length > 180) s = s.substring(0, 180);
  return s.isEmpty ? 'file' : s;
}

/// رفع ملفّ إلى دلو Nebras (مثل `smartUpload` → Storage). الترتيب الثابت:
/// Storage أولاً ثم نُعيد رابط التنزيل ليُكتب في Firestore.
///
/// [onProgress] نسبة مئويّة 0..100. [isAborted] لإلغاء الرفع.
Future<UploadResult> smartUpload({
  required Uint8List bytes,
  required String filename,
  String contentType = 'application/octet-stream',
  String folder = 'dashboard/uploads',
  void Function(double pct)? onProgress,
  bool Function()? isAborted,
  void Function(UploadTask task)? onTaskCreated,
}) async {
  final storage = FirebaseStorage.instance;
  final safeName = _sanitizeSegment(filename);
  final cleanFolder = folder.replaceAll(RegExp(r'^/+|/+$'), '');
  final path =
      '$cleanFolder/${DateTime.now().millisecondsSinceEpoch}_$safeName';
  final ref = storage.ref(path);
  final ct =
      contentType.trim().isEmpty ? 'application/octet-stream' : contentType;

  final task = ref.putData(bytes, SettableMetadata(contentType: ct));
  onTaskCreated?.call(task);

  task.snapshotEvents.listen((snap) {
    if (isAborted != null && isAborted()) {
      task.cancel();
      return;
    }
    final total = snap.totalBytes <= 0 ? bytes.length : snap.totalBytes;
    onProgress?.call((snap.bytesTransferred / total) * 100);
  });

  try {
    await task;
  } on FirebaseException catch (e) {
    if (e.code == 'canceled') {
      throw Exception('تمّ إلغاء الرفع.');
    }
    rethrow;
  }

  if (isAborted != null && isAborted()) {
    throw Exception('تمّ إلغاء الرفع.');
  }

  final url = await ref.getDownloadURL();
  return UploadResult(
    url: url,
    path: path,
    contentType: ct,
    size: bytes.length,
    filename: safeName,
  );
}

/// رفع ملفّ من المسار (بثّ — لا يُحمَّل كاملاً في الذاكرة). يُستعمل للملفّات
/// الكبيرة في الرفع المتعدّد لتفادي OOM.
Future<UploadResult> smartUploadFile({
  required File file,
  required String filename,
  String contentType = 'application/octet-stream',
  String folder = 'dashboard/uploads',
  void Function(double pct)? onProgress,
  bool Function()? isAborted,
  void Function(UploadTask task)? onTaskCreated,
}) async {
  final storage = FirebaseStorage.instance;
  final safeName = _sanitizeSegment(filename);
  final cleanFolder = folder.replaceAll(RegExp(r'^/+|/+$'), '');
  final path =
      '$cleanFolder/${DateTime.now().millisecondsSinceEpoch}_$safeName';
  final ref = storage.ref(path);
  final ct =
      contentType.trim().isEmpty ? 'application/octet-stream' : contentType;
  final size = await file.length();

  final task = ref.putFile(file, SettableMetadata(contentType: ct));
  onTaskCreated?.call(task);

  task.snapshotEvents.listen((snap) {
    if (isAborted != null && isAborted()) {
      task.cancel();
      return;
    }
    final total = snap.totalBytes <= 0 ? size : snap.totalBytes;
    onProgress?.call((snap.bytesTransferred / total) * 100);
  });

  try {
    await task;
  } on FirebaseException catch (e) {
    if (e.code == 'canceled') {
      throw Exception('تمّ إلغاء الرفع.');
    }
    rethrow;
  }

  if (isAborted != null && isAborted()) {
    throw Exception('تمّ إلغاء الرفع.');
  }

  final url = await ref.getDownloadURL();
  return UploadResult(
    url: url,
    path: path,
    contentType: ct,
    size: size,
    filename: safeName,
  );
}
